cli/deduplifhirLib/normalization.py

```diff
@@ -0,0 +1,337 @@
+    """
+    Add permutations of the keys from a dict and map them to the same value
+    as the original key. 
+
+    For example we make every key also match to its lowercase version,
+    as well as mapping to the version with punctuation and proper spaces. 
+
+
+    Arguments:
+        symbol_dict: dictionary of symbols to add permutations to 
+    
+    Returns:
+        Expanded version of dict input
+    """
+    #Add some permutations to the dictionary manually once
+    abvs = list(symbol_dict.keys())
+    for key in abvs:
+        symbol_dict[key.lower()] = symbol_dict[key]
+        symbol_dict[key[:-1] + ','] = symbol_dict[key]
+        symbol_dict[key[:-1] + '.'] = symbol_dict[key]
+
+
+        if key[-1] == '.':
+            symbol_dict[key[:-2] + ' '] = symbol_dict[key]
+            symbol_dict[key[:-2] + ', '] = symbol_dict[key]
+
+    return symbol_dict
+
+
+def compile_abbreviation_map_regex(symbol_dict):
+    """
+    Compile a regular expression that converts the dictionary pattern into a 
+    regex that matches and replaces every instance of the keys found in the dict
+    with their proper normalized values.
+
+
+    Arguments:
+        symbol_dict: dictionary of symbols to replace
+    
+    Returns:
+        Regex pattern object.
+    """
+
+    # Create a regex REPLACE_PLACE_NAME_SYMBOLS_PATTERN that matches any of the
+    #  keys in the abbreviation dictionary
+    return re.compile('|'.join(re.escape(key) for key in symbol_dict.keys()))
+
+REPLACE_PLACE_NAME_SYMBOLS_PATTERN = compile_abbreviation_map_regex(
+    add_permutations_to_dict_pattern(
+        PLACE_ABBREVIATION_SYMBOLS
+    )
+)
+
+REPLACE_PROPER_NAME_SYMBOLS_PATTERN = compile_abbreviation_map_regex(
+    add_permutations_to_dict_pattern(
+        NAME_ABBREVIATION_SYMBOLS
+    )
+)
+
+
+def replace_abbreviations(input_text,pattern=REPLACE_PLACE_NAME_SYMBOLS_PATTERN,symbols=PLACE_ABBREVIATION_SYMBOLS):
+    """
+    Normalizes common abbreviations with a pre-compiled regular expression.
+
+    Arguments:
+        input_text: the input_text to remove the abbreviations from.
+
+    Returns:
+        The input string without the abbreviations.
+    """
+    # Define a function to use as the replacement argument in re.sub
+    def replacer(match):
+        return " " + symbols[match.group(0)] + " "
+
+    # Use re.sub with the REPLACE_PLACE_NAME_SYMBOLS_PATTERN
+    #  and replacer function to replace abbreviations
+    return pattern.sub(replacer, input_text)
+
+
+def remove_punctuation(input_text):
+    """
+    Removes punctuation from the given string.
+
+    Arguments:
+        input_text: the input_text to remove the punctuation from
+    
+
+    Returns:
+        The input string without punctuation
+    """
+    symbols_to_remove = ".,?!;:\'\""
+    text_copy = input_text
+
+    for symbol in symbols_to_remove:
+        text_copy = text_copy.replace(symbol,'')
+    return text_copy
+
+
+def british_to_american(input_text):
```

Comment:
That's fair. I was mainly doing it for stuff like there being a 'Centre' instead of a 'Center' but on hindsight it might be better to just leave that alone.